### PR TITLE
Refactor/Refactor controller logic

### PR DESCRIPTION
**Pull request by iGxnon**
Base on #19

Please briefly answer these questions:

* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)

    This PR refactor controller into three parts: `Extractor`, `Factory`, `Manager`.
    Introducing these components benefits us by segregating the logic of the controller, facilitating unit testing, and preparing for the eventual replacement of the StatefulSet controller.

* what changes does this pull request make?

    As above.

* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)

    No

### CODE REVIEW

File: operator-k8s/src/manager/cluster.rs [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+    /// A backup job
+    BackupJob,
+}
+
+impl Component {
+    /// Get the component name
+    fn label(&self) -> &str {
+        match *self {
+            Component::Node => "node",
+            Component::Service => "srv",
```
[the remark this answers is not in the document]
Comment (iGxnon) [reply operator-k8s/src/manager/cluster.rs]: done

File: operator-k8s/src/manager/cluster.rs [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+            Component::BackupJob => "job",
```
[the remark this answers is not in the document]
Comment (iGxnon) [reply operator-k8s/src/manager/cluster.rs]: done, it is called backup now

**Review by Phoenix500526 — APPROVED** [review]
LGTM

File: operator-k8s/src/operator.rs [hunk 3]
```diff
@@ -62,73 +62,159 @@ impl Operator {
...
+        });
+        #[allow(unsafe_code)] // safe
+        let _ig = tokio::spawn(async move {
+            let mut shutdown = graceful_shutdown;
+
+            {
+                // Safety:
+                // Some hacking to make future generated from `graceful_shutdown` to be 'static
+                // The 'static marker is required by `kube::runtime::Controller::graceful_shutdown_on`
+                // and it is not good for our design.
```
[the remark this answers is not in the document]
Comment (iGxnon) [reply operator-k8s/src/operator.rs:134]: I think we have to wait for `kube-rs` to introduce a new shutdown api or change the constraints(might be a long term). There's nothing wrong with this, so shouldn't be any need to refactor

**Review by bsbds — APPROVED** [review]

**Pull request merged**